Adding a couple of new Samsung browser versions, new devices

jaydiablo opened the pull request:
This fixes #1285.

Added Samsung Browser 4.2 and 5.2.

Added new Samsung devices.

Consolidated the patterns/devices in the Samsung browser file.

mimmi20 commented on resources/user-agents/browsers/samsung/samsung-android-browser.json:
```diff
@@ -1,6 +1,6 @@
 {
   "division": "Samsung Browser #MAJORVER#.#MINORVER#",
-  "versions": [5, 4, "3.5", "3.4", "3.3", "3.2", "3.1", 3, "2.1", 2],
+  "versions": ["5.2", 5, "4.2", 4, "3.5", "3.4", "3.3", "3.2", "3.1", 3, "2.1", 2],
```
Please add the versions 4.1 and 5.1 too.

jaydiablo commented:
I don't believe 5.1 and 4.1 exist.  I just did a search through our server logs for "SamsungBrowser/4.1" and "SamsungBrowser/5.1" and found nothing. 

I also looked here: https://html5test.com/results/search.html?query=samsung%20internet also no 4.1 or 5.1.

Those versions aren't listed on Samsung's Release Notes page for developers: http://developer.samsung.com/internet/android/releases (but neither is 5.2).

I also found this APK Mirror site that allows you to download older versions of Android APK files: https://www.apkmirror.com/apk/samsung-electronics-co-ltd/samsung-internet-for-android/ (also no 4.1 and 5.1)

jaydiablo commented:
FWIW, I just searched through our logs (past 7 years or so) and extracted all of the Samsung Browser versions that we've seen when using this pattern as the search:

`Mozilla/5.0 (*Android* Build/*) applewebkit* (*khtml*like*gecko*) SamsungBrowser/*.*Chrome*Safari*`

Which is similar to the patterns in the  samsung-android-browser.json file, I just replaced `#PLATFORM#` with `*Android*`.

From about 5.5 million hits that matched that pattern here are the versions we've seen (ordered by number of hits):
	
`4.0`, `2.1`, `3.3`, `3.2`, `5.0`, `3.4`, `3.5`, `3.0`, `4.2`, `2.0`, `5.2`

I'd recommend that these are the only versions in the file, as it doesn't seem like Samsung has released anything else yet.

jaydiablo commented:
I also grepped through the list of useragents that I've used in the past that come from other parsers' test suites (47k different useragents), and none of them have Samsung Browser 3.1, 4.1 or 5.1.

mimmi20 commented:
Not existing versions or combinations are a issue of browscap. There are a lot of pattern in browscap (combination of platform version, browser version and devics) which does not exist.

jaydiablo commented:
@mimmi20 Yeah, and I think that's mainly the result of the way that the version expansion is done (if you add a version to a file, every device/platform combination in that file gets that version as well). I'm not sure that the alternative is better though (trying to specify the version(s) as a property of the child and just expanding the version for that child's platforms/devices, as this would require a lot more work to keep everything up to date).

That said, I think it would be nice to not include versions that don't exist. Can be helpful for weeding out the fakes. Speaking of, should tests be added to make sure that the 5.1, 4.1 and 3.1 versions get detected as "Default Browser" since it seems pretty obvious that they don't exist (and to prevent them from being added in the future)?
